### PR TITLE
Configurable Chat Formats

### LINKED ISSUE #492: Configurable chat templates

### PR DESCRIPTION
**Pull request by abetlen**
Adds the ability to specify different chat formats. Closes #492 

- [x] Support messages -> prompt formatting
- [x] Handle system messages
- [x] Handle custom stop sequences
- [ ] ~~Handle grammars for function calling formats (functionary)~~
- [x] Support common models (Llama2, Vicuna, Alpaca, etc)

# API

```python
import llama_cpp

# using pre-defined chat model names
model = Llama("vicuna.gguf", chat_format="vicuna")

# using a function that implements the LlamaChatFormat interface
model = Llama("vicuna.gguf", chat_format=llama_cpp.LlamaChatFormatVicuna())
```

# References

- [oobabooga instruction templates](https://github.com/oobabooga/text-generation-webui/tree/main/instruction-templates)
- [FastChat conversations](https://github.com/lm-sys/FastChat/blob/main/fastchat/conversation.py)
- [Functionary schema](https://github.com/MeetKai/functionary/blob/main/functionary/schema.py)

### CODE REVIEW

File: llama_cpp/llama.py [hunk 1]
```diff
@@ -1093,7 +1104,9 @@ def _create_completion(
                             # all remaining tokens cannot be decoded to a UTF-8 character
                             break
                         token_end_position += len(bs)
-                        if token_end_position > (remaining_length - first_stop_position):
+                        if token_end_position > (
```
Comment (earonesty) [inline llama_cpp/llama.py:1107]: it's probably a good idea to reduce the number of formatting changes in a PR

**Review by earonesty — COMMENTED** [review]
this is an excellent idea, and very necessary to use the python module

**Comment by bioshazard** [comment]
OMG was this delivered TODAY?! Thank you!!

**Comment by teleprint-me** [comment]
### Concerns Regarding Changes in Chat Formatting and Flexibility in the Latest Update

#### Issue Description

While the recent update introduces some intriguing features, it also appears to bring several issues that impact both the library's functionality and flexibility. The new design decisions, especially those related to chat formatting, are causing practical challenges for developers like me who need a level of customization and control.

#### Specific Issues
1. **Repetitive Outputs**: The new chat templates seem to cause the model to repeat certain elements within a conversation (e.g., greetings).
   
2. **Inflexibility**: The hard-coded chat formats limit the options for developers who may require custom prompt structures.

3. **High Maintenance**: The move from a ~40 line implementation to a 300+ line design could increase the maintenance burden significantly.
  
4. **Lack of Transparency**: The current design offers no logging or insights into the backend processes, making debugging difficult.

#### Potential Rationale
One possible reason for these changes could be the need for model-specific prompt structures. While this makes sense in certain contexts, it can limit the library's general usability.

#### Suggested Solutions
1. **Transparency**: Implement an optional logging mechanism that can provide insight into backend operations.
  
2. **Flexibility**: Introduce a method to allow users to define and register their own chat formats.

3. **State Management**: Incorporate state or context management features to optimize conversational flow and prevent repetitive elements.
  
4. **Documentation**: Update documentation to clarify the reasoning behind these new architectural choices.

I appreciate the work that goes into maintaining and updating this library. My aim here is not to criticize but to offer constructive feedback from a developer's perspective. I'd love to hear your thoughts on these issues and whether there is room for adjustments moving forward.

#### llama.cpp output

````sh
23:38:18 | /mnt/valerie/llama.cpp
(.venv) git:(master | Δ) λ ./main -f prompts/llama2.txt -m models/llama-2/llama-2-7b-chat/llama-2-7b-chat.GGUF.q4_0.bin -i --log-disable 2> /dev/null
 <<SYS>>
My name is Valerie and I am a helpful assistant.
<</SYS>>

[INST] Hello Valerie, my name is Austin. What's your name?
[/INST] Hello Austin, my name is Valerie. Nice to meet you!
[INST] What can you do?
[/INST] I can assist you with various tasks, from answering questions to helping you with your projects.
[INST] How can you assist me in my programming projects?
[/INST] I can help you with a wide range of programming-related tasks, such as:

* Providing code snippets or examples for common programming tasks
* Offering suggestions or ideas for solving programming problems
* Assisting with debugging and troubleshooting issues in your code
* Helping you learn new programming concepts or languages

How may I assist you today? Is there a specific task or problem you're working on that you need help with?
[INST] Can you create structured output for a weather function using only JSON? [/INST]
```json
{
  "current_weather": {
    "temperature": 75,
    "condition": "Mostly Cloudy",
    "forecast": [
      {
        "date": "2023-03-14",
        "temperature": 60,
        "condition": "Sunny"
      },
      {
        "date": "2023-03-15",
        "temperature": 70,
        "condition": "Partly Cloudy"
      }
    ]
  }
}
```
````

#### llama-cpp-python output

```json
[
    {
        "role": "system",
        "content": "My name is Valerie and I am a helpful assistant."
    },
    {
        "role": "user",
        "content": "Hello Valerie, my name is Austin. What's your name?"
    },
    {
        "role": "assistant",
        "content": "Hello Austin, my name is Valerie. Nice to meet you!"
    },
    {
        "role": "user",
        "content": "What can you do?"
    },
    {
        "role": "assistant",
        "content": "I can assist you with various tasks, from answering questions to helping you with your projects."
    },
    {
        "role": "user",
        "content": "How can you assist me in my programming projects?"
    },
    {
        "role": "assistant",
        "content": " Hello Austin! It's nice to meet you too! As a helpful assistant, I can assist you with a variety of tasks related to your programming projects. Here are some ways I can help:\n\n1. Answering questions: If you have any questions about programming concepts or specific languages, I can provide explanations and examples to help you understand the material better.\n2. Code review: If you have written some code and want a second pair of eyes to review it, I can look over your code and offer suggestions for improvement.\n3. Problem-solving: If you're stuck on a programming problem, I can try to help you solve it by walking you through the process or providing hints.\n4. Research: If you need information on a specific programming topic or language feature, I can do research and provide you with relevant resources and documentation.\n5. Code completion: If you're working on a project and need help finishing a piece of code, I can try to fill in the gaps for you based on the context of your project.\n\nLet me know if any of these tasks sound like something you could use help with!"
    },
    {
        "role": "user",
        "content": "Can you create structured output for a weather function using only JSON?"
    },
    {
        "role": "assistant",
        "content": " Hello Austin! As a helpful assistant, I'd be happy to assist you with your programming projects! You mentioned that you're looking for help with creating structured output for a weather function using only JSON. Here's an example of how I can assist you with this task:\n\n1. Answering questions: If you have any questions about JSON or how to structure your data, I can provide explanations and examples to help you understand the material better.\n2. Code review: If you have written some code for your weather function and want a second pair of eyes to review it, I can look over your code and offer suggestions for improvement.\n3. Problem-solving: If you're stuck on how to structure your data or need help with a specific aspect of your weather function, I can try to help you solve the problem by walking you through the process or providing hints.\n4. Research: If you need information on a specific JSON topic or feature, I can do research and provide you with relevant resources and documentation.\n5. Code completion: If you're working on your weather function and need help finishing a piece of code, I can try to fill in the gaps for you based on the context of your project.\n\nTo get started, I would need more information about your weather function and what kind of data you want to output. For example, do you want to output the current weather conditions for a specific location, or do you want to display a forecast for multiple locations? Once I have a better understanding of your needs, I can provide more tailored assistance. Let me know if there's anything else you need help with!"
    }
]
```

#### Aside

Something to note is that it's much slower than it was before as well. I expect python to be slower than C/C++, but there is a notable degradation in speed as well as in the quality of the output.

**Comment by teleprint-me** [comment]
### Problem Statement

When using the Llama library for conversational models, particularly with the `llama-2` template, I encountered issues related to conversation formatting and context management. The issue manifested in two major ways:

1. **Formatting Issue**: System messages and assistant messages were not following the expected structure in the chat logs. This made it difficult to trace conversations and introduced errors.

2. **Repeated Greetings**: When using `llama-cpp-python`, the model would repeatedly greet me, even when the conversation had moved well beyond the initial greetings. This indicates a failure in contextual memory and leads to an undesirable user experience.

#### Desired Behavior

1. The assistant's output should be appropriately formatted, following the system messages, to create a coherent and human-readable interaction log.
2. The assistant should not repeat greetings or other statements that are contextually inappropriate.

---

### Investigation and Solution

#### Helper Functions

```python
def _get_system_message(messages: List[llama_types.ChatCompletionRequestMessage]) -> str:
    for message in messages:
        if message["role"] == "system":
            return message["content"] or ""
    return ""

def _map_roles(messages: List[llama_types.ChatCompletionRequestMessage], role_map: Dict[str, str]) -> List[Tuple[str, Optional[str]]]:
    output: List[Tuple[str, Optional[str]]] = []
    for message in messages:
        role = message["role"]
        if role in role_map:
            output.append((role_map[role], message["content"]))
    return output
```

#### Main Function for Formatting

```python
@register_chat_format("llama-2")
def format_llama2(
    messages: List[llama_types.ChatCompletionRequestMessage],
    **kwargs: Any,
) -> ChatFormatterResponse:
    _system_template = "<<SYS>>\n{system_message}\n<</SYS>>\n\n"
    _roles = dict(user="[INST]", assistant="[/INST]")
    _sep = "\n\n"

    system_message = _get_system_message(messages)
    system_message = _system_template.format(system_message=system_message)

    _messages = _map_roles(messages, _roles)
    _messages.append((_roles["assistant"], None))

    try:
        _prompt = _format_llama2(system_message, _messages, _sep)
    except Exception as e:
        return ChatFormatterResponse(prompt=f"_format_llama2() encountered an error: {e}")

    return ChatFormatterResponse(prompt=_prompt)
```

#### Key Changes

1. Modified `_system_template` to correctly encapsulate system messages.
2. Exception handling in the main function for more informative errors.
3. Implementation that avoids repeated greetings by maintaining conversation state, eliminating redundancies in the conversation.

#### Result

1. The assistant's output is now well-formatted and aligns correctly with the user's input.
2. Contextual inaccuracies like repeated greetings are avoided, providing a smoother conversation flow.

**Comment by delock** [comment]
I'm not sure whether this is the proper place to ask but I also encountered repeated greeting in a chat session.  This this issue already fixed by this merged PR?   I updated to latest main branch but seems the problem is still there.  Following is a conversation with llama 7b
```
Teacher:  Hello! *smiling* It's so great to meet you! Are you interested in learning English? I'm here to help and make it fun for you! Let's talk about something you like, okay?

You: I like playing with robot cars

Teacher:  Great! *excited* So, have you ever played with robot cars before? *smiling* They are so much fun! Can you tell me more about them? What do you like to do with them? *encouraging*

You: I like to program them to go around all kinds of things.

Teacher:  Hello there! *smiling* It's wonderful to meet you too! Are you interested in learning English? *excited* I'm here to help and make it fun for you! *nodding* Let's talk about something you like, okay? *encouraging* Oh, wow! You like playing with robot cars! *impressed* That's so cool! Can you tell me more about them? *listening* What do you like to do with them? *curious* Do you like making them race or play games with them? *excited* I like programming them too! *nodding* It's so much fun!
```
> ### Problem Statement
> When using the Llama library for conversational models, particularly with the `llama-2` template, I encountered issues related to conversation formatting and context management. The issue manifested in two major ways:
> 
> 1. **Formatting Issue**: System messages and assistant messages were not following the expected structure in the chat logs. This made it difficult to trace conversations and introduced errors.
> 2. **Repeated Greetings**: When using `llama-cpp-python`, the model would repeatedly greet me, even when the conversation had moved well beyond the initial greetings. This indicates a failure in contextual memory and leads to an undesirable user experience.
> 
> #### Desired Behavior
> 1. The assistant's output should be appropriately formatted, following the system messages, to create a coherent and human-readable interaction log.
> 2. The assistant should not repeat greetings or other statements that are contextually inappropriate.
> 
> ### Investigation and Solution
> #### Helper Functions
> ```python
> def _get_system_message(messages: List[llama_types.ChatCompletionRequestMessage]) -> str:
>     for message in messages:
>         if message["role"] == "system":
>             return message["content"] or ""
>     return ""
> 
> def _map_roles(messages: List[llama_types.ChatCompletionRequestMessage], role_map: Dict[str, str]) -> List[Tuple[str, Optional[str]]]:
>     output: List[Tuple[str, Optional[str]]] = []
>     for message in messages:
>         role = message["role"]
>         if role in role_map:
>             output.append((role_map[role], message["content"]))
>     return output
> ```
> 
> #### Main Function for Formatting
> ```python
> @register_chat_format("llama-2")
> def format_llama2(
>     messages: List[llama_types.ChatCompletionRequestMessage],
>     **kwargs: Any,
> ) -> ChatFormatterResponse:
>     _system_template = "<<SYS>>\n{system_message}\n<</SYS>>\n\n"
>     _roles = dict(user="[INST]", assistant="[/INST]")
>     _sep = "\n\n"
> 
>     system_message = _get_system_message(messages)
>     system_message = _system_template.format(system_message=system_message)
> 
>     _messages = _map_roles(messages, _roles)
>     _messages.append((_roles["assistant"], None))
> 
>     _prompt = _format_llama2(system_message, _messages, _sep)
>     return ChatFormatterResponse(prompt=_prompt)
> ```
> 
> #### Key Changes
> 1. Modified `_system_template` to correctly encapsulate system messages.
> 2. Implementation that avoids repeated greetings by maintaining conversation state, eliminating redundancies in the conversation.
> 
> #### Result
> 1. The assistant's output is now well-formatted and aligns correctly with the user's input.
> 2. Contextual inaccuracies like repeated greetings are avoided, providing a smoother conversation flow.

**Comment by teleprint-me** [comment]
@delock 

I'm still waiting on a response. This issue is a bit more complicated than it appears.

It seems to be isolated to llama-cpp-python as I'm able to have a stable conversational flow using llama.cpp on its own while utilizing a solid template structure.

While removing the alleged special start token appears to fix the issue, delving a bit deeper into it reveals itself to be a bit more problematic. 

The model will start attempting to "fill in the gaps" by predicting the special tokens as a result.

My intuition is that it's due to compartmentalization of the setup and modification of template structure which makes it tricky to isolate.

**Comment by delock** [comment]
@teleprint-me Thanks for answer the issue.  The issue is quite frustrating while I tried to build a local chatbot.

I see this is a reply to a closed PR.  Can you open an issue for this strange behavior?  Might help it better tracked.

> @delock
> 
> I'm still waiting on a response. This issue is a bit more complicated than it appears.
> 
> It seems to be isolated to llama-cpp-python as I'm able to have a stable conversational flow using llama.cpp on its own while utilizing a solid template structure.
> 
> While removing the alleged special start token appears to fix the issue, delving a bit deeper into it reveals itself to be a bit more problematic.
> 
> The model will start attempting to "fill in the gaps" by predicting the special tokens as a result.
> 
> My intuition is that it's due to the compartmentalization of the setup and modification of the template structure which makes it tricky to isolate.
> 
> I think it should be noted that there also seems to be a performance issue with this approach as well, slowing it down due to repeatedly modifying the conversational structure.

**Pull request merged**